Update departures and climatology APIs with reference period

tomvothecoder opened the pull request:
## Description
<!--
  Please include a summary of the change and which issue is fixed.
  Please also include relevant motivation and context.
  List any dependencies that are required for this change.
-->

- Closes #387 
- Add `reference_period` parameter to `climatology()` and `departures()` to subset climatology 
- In `departures()`, calculate the the grouped average of the observation data before calculating anomalies
  - This behavior makes `departures()` align with CDAT's APIs for calculating departures
  - If the input frequency from the dataset is the same as the output frequency `freq` arg, no grouped average is performed since it is unnecessary and incurs a performance cost 
- Update `departures()` notebook with `reference_period` example
 
## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

If applicable:

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass with my changes (locally and CI/CD build)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have noted that this is a breaking change for a major release (fix or feature that would cause existing functionality to not work as expected)

tomvothecoder reviewed (commented):
Hi @lee1043 and @pochedls, this PR for implementing `reference_period` in `climatology()` and `departures()` is ready for review.

I decided to not implement reference period subsetting for `average()` and `group_average()` because the user can easily do that themselves before calling the temporal APIs, e.g., `ds.isel(time=slice(....))`.

### Validation
In addition to the updated unit tests, below is the validation script that compares method 1 (@lee1043's [cookbook](https://github.com/xCDAT/xcdat_test/blob/main/cookbook/seasonal_departure_from_baseline_clim.ipynb)) to method 2 (this PR implementation) for calculating climatologies and departures relative to a reference period. The result of the script shows that **both methods produce identical results.**

```python
# flake8: noqa F401 E256
#%%
import matplotlib.pyplot as plt

import xcdat as xc

#%%
infile = "./input/HadISST_sst.nc"
data_var = "sst"  #  name of data variable

ds = xc.open_dataset(infile)
ds[data_var] = ds[data_var].where(ds[data_var] >= -0.5)  # exclude missing value

#%%
# =======================================
#  METHOD 1 - xCDAT + xarray
# =======================================
# 1. Get seasonal average of the full time series
ds_season = ds.temporal.group_average(data_var, freq="season")
ds_season

#%%
# 2. Temporal subset of original dataset to get reference climatology
ds_reference = ds.sel(time=slice("1971-01-01", "2000-12-31"))
ds_clim1 = ds_reference.temporal.climatology(data_var, freq="season")
ds_clim1[data_var].isel(time=0).plot(cmap="viridis")

#%%
# 3. Calculate departures
ds_departs1 = ds_season.copy()
ds_departs1[data_var + "_departure"] = ds_departs1[data_var].groupby(
    "time.season"
) - ds_clim1[data_var].groupby("time.season").mean("time")
ds_departs1[data_var + "_departure"].isel(time=0).plot()

# %%
# =======================================
#  METHOD 2 - xCDAT
# =======================================
# 1. Temporal subset of original dataset to get reference climatology
ds_clim2 = ds.temporal.climatology(
    data_var, freq="season", reference_period=("1971-01-01", "2000-12-31")
)

ds_clim2[data_var].isel(time=0).plot(cmap="viridis")

# Check equals method 1
ds_clim2["sst"].equals(ds_clim2["sst"])  # True

#%%
# 2. Calculate departures
ds_departs2 = ds.temporal.departures(
    data_var, freq="season", reference_period=("1971-01-01", "2000-12-31")
)
ds_departs2[data_var].isel(time=0).plot()

#%%
# =======================================
#  Compare Outputs
# =======================================
# Drop season sub-coordinates
ds_departs1_new = ds_departs1.drop_vars("season")

# Get the expected and resulting departures
method1_anom = ds_departs1_new["sst_departure"].rename("sst")
method2_anom = ds_departs2["sst"]
#%%
# Compare equality
method2_anom.equals(method1_anom)  # True

#%%
# Compare identical (drop attributes)
expected2 = method1_anom.copy()
result2 = method2_anom.copy()

expected2.attrs = {}
result2.attrs = {}

expected2.identical(result2)  # Truegithub
```

pochedls commented on xcdat/temporal.py:
```diff
@@ -375,6 +384,10 @@ def climatology(
         keep_weights : bool, optional
             If calculating averages using weights, keep the weights in the
             final dataset output, by default False.
+        reference_period : Optional[Tuple[str, str]], optional
+            The climatological reference period, which is a subset of the entire
+            time series. This parameter accepts a tuple of strings in the format
+            'yyyy-mm-dd'. For example, ``('1850-01-01', 1899-12-31')``.
```
```suggestion
        reference_period : Optional[Tuple[str, str]], optional
            The climatological reference period, which is a subset of the entire
            time series. This parameter accepts a tuple of strings in the format
            'yyyy-mm-dd'. For example, ``('1850-01-01', 1899-12-31')``. If no
            value is provided, the climatological reference period will be the
            full period covered by the dataset.
```

Simply adds: "If no value is provided, the climatological reference period will be the full period covered by the dataset."

pochedls commented on xcdat/temporal.py:
```diff
@@ -539,6 +548,11 @@ def departures(
         keep_weights : bool, optional
             If calculating averages using weights, keep the weights in the
             final dataset output, by default False.
+        reference_period : Optional[Tuple[str, str]], optional
+            The climatological reference period, which is a subset of the entire
+            time series and used for calculating departures. This parameter
+            accepts a tuple of strings in the format 'yyyy-mm-dd'. For example,
+            ``('1850-01-01', 1899-12-31')``.
```
```suggestion
        reference_period : Optional[Tuple[str, str]], optional
            The climatological reference period, which is a subset of the entire
            time series and used for calculating departures. This parameter
            accepts a tuple of strings in the format 'yyyy-mm-dd'. For example,
            ``('1850-01-01', 1899-12-31')``. If no value is provided, the 
            climatological reference period will be the full period covered by 
            the dataset.
```

Simply adds: "If no value is provided, the climatological reference period will be the full period covered by the dataset."

pochedls commented on xcdat/temporal.py:
```diff
@@ -611,48 +632,59 @@ def departures(
 
-        # Group the observation data variable.
-        dv_obs = _get_data_var(ds, data_var)
+        # 1. Get the grouped average of the observation data variable.
+        ds_avg = ds.temporal.group_average(
+            data_var, self._freq, weighted, keep_weights, season_config=season_config
+        )
+        dv_obs = ds_avg[data_var].copy()
+
+        # 2. Group the observation data variable by the departures frequency.
+        # This step allows us to perform xarray's grouped arithmetic to calculate
+        # departures.
         self._labeled_time = self._label_time_coords(dv_obs[self.dim])
         dv_obs_grouped = self._group_data(dv_obs)
```
Thanks for catching and fixing this. But what if the input/output frequencies are the same? Should we still call `group_average()`?

pochedls reviewed (commented):
Overall, this PR seems to work great and makes sense to me. It takes me a long time to re-orient myself to the inner workings of the `.temporal` accessor so I'm hoping @lee1043 can also give this a careful review. 

* One thing I noticed is that calling `ds.temporal.departures` drops `time_bnds` – which I don't think needs to happen (I assume xarray doesn't know how to deal with this during the subtraction of the climatology). I think the time bounds can be added back in after subtraction. I noticed this in verifying that this PR was accurate (I get accuracy of order 10<sup>-4</sup> – 10<sup>-7</sup>, I think because I added time bounds back in using `add_missing_bounds`, which is problematic). 
* Note to others that I did need to update my dev environment to pass the unit tests
* Made some minor documentation suggestions
* Steps zero and one of `departures` call `_set_arg_attrs()`, `.copy()`, `_preprocess_dataset()`, and `group_average()`. This overlaps strongly with `_averager()` – maybe things could be streamlined here (though I see that some of the order-of-operations is different, which may be why they are different. 
* As noted, departures calls `group_average()` on each call – I think this may not be necesary if the input/output frequency are the same?

tomvothecoder commented on xcdat/temporal.py:
```diff
@@ -669,15 +671,38 @@ def departures(
+            # and therefore dropped.
             ds_avg = ds_avg.drop_vars(self._labeled_time.name)
 
         if weighted and keep_weights:
```
In `departures()`, I updated the order of operations and added clearer comments to outline the steps and why they need to be done this specific way.

tomvothecoder commented:
>  * One thing I noticed is that calling `ds.temporal.departures` drops `time_bnds` – which I don't think needs to happen (I assume xarray doesn't know how to deal with this during the subtraction of the climatology). I think the time bounds can be added back in after subtraction. I noticed this in verifying that this PR was accurate (I get accuracy of order 10-4 – 10-7, I think because I added time bounds back in using `add_missing_bounds`, which is problematic).

Before this PR:
- `departures()` used to maintain `time_bnds` because the output frequency was the same as the input frequency (no `group_average()` call).

In this PR:
- Since we now call `group_average()`, `time_bnds` are dropped because the `time` coordinates are grouped which changes the `time` frequency and shape.

Solutions/Workarounds:
1. As you mentioned, get new `time_bnds` for the new grouped `time` coordinates using `add_missing_bounds()`. The accuracy _should_ improve with #412.
2. Keep a separate `time_original` dimension/coordinates to keep `time_bnds` in the dataset (otherwise we get `time_bnds` with `NaT` values since `time` frequency and shape changes).

---

>  * Steps zero and one of `departures` call `_set_arg_attrs()`, `.copy()`, `_preprocess_dataset()`, and `group_average()`. This overlaps strongly with `_averager()` – maybe things could be streamlined here (though I see that some of the order-of-operations is different, which may be why they are different.

I saw your point and played around with different implements using `_average()` to tried to find ways to streamline the implementation. I wasn't able to find a simple way to call `self._average()` without making significant modifications to how the `TemporalAccessor` class attributes are set (e.g., `self._mode`). **We can probably try refactoring this in the future if we wanted to.**

I updated the order-of-operations and made clearer comments in the `departures()` method so it is hopefully easier to understand. Please check out this [commit](https://github.com/xCDAT/xcdat/pull/417/commits/25832ddf5357bf043ecbd300aa03e71308840987) diff and let me know if those comments make sense.

---

>   * As noted, departures calls `group_average()` on each call – I think this may not be necesary if the input/output frequency are the same?

I also don't think we need to perform group averaging if the input/output frequencies are the same. **It still works either way** (and converts time coordinates to `cftime` if they aren't already), but it might improve performance by removing unnecessary operations.

We can try determining if `group_average()` is necessary by using the existing `_infer_freq()` method and comparing the result to the input `freq` argument.

tomvothecoder commented on xcdat/temporal.py:
```diff
@@ -611,48 +639,88 @@ def departures(
-        dv_obs_grouped = self._group_data(dv_obs)
+        # 3. Calculate the grouped average and climatology of the data variable.
+        # ----------------------------------------------------------------------
+        # The climatology and grouped average APIs are called on the copied
+        # dataset to create separate instances of the `TemporalAccessor` class.
+        # This is done to avoid overriding the attributes of the current
+        # instance of `TemporalAccessor` (set in step #1 above).
```
If the `departures()` performance or memory usage is ever an issue, we can look into refactoring these steps to avoid copying. Copying datasets multiple times might not be an issue though because [xr.Dataset.copy()](https://docs.xarray.dev/en/stable/generated/xarray.Dataset.copy.html) defaults to `deep=False` (_"... a shallow copy of each of the component variable is made, so that the underlying memory region of the new dataset is the same as in the original dataset."_)

On the other hand, [xr.DataArray.copy()](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.copy.html) defaults to `deep=True`.

tomvothecoder commented on xcdat/temporal.py:
```diff
@@ -611,48 +639,88 @@ def departures(
+        # This is done to avoid overriding the attributes of the current
+        # instance of `TemporalAccessor` (set in step #1 above).
+
+        # Group averaging is only required if the dataset's frequency (input)
+        # differs from the `freq` arg (output).
+        ds_obs = ds.copy()
+        inferred_freq = self._infer_freq()
+        if inferred_freq != freq:
+            ds_obs = ds_obs.temporal.group_average(
+                data_var,
+                freq,
+                weighted,
+                keep_weights,
+                season_config,
+            )
```
Added conditional for group averaging only if the input and output frequencies are the same. This should save on performance costs for this case.

Related to https://github.com/xCDAT/xcdat/pull/417#discussion_r1113628929 and https://github.com/xCDAT/xcdat/pull/417#pullrequestreview-1308232306.
